pkg/apis/jenkins.io/v1/types_fact.go

```diff
@@ -0,0 +1,99 @@
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:noStatus
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:openapi-gen=true
+
+// Fact represents observed facts
+type Fact struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	Spec   FactSpec   `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+	Status FactStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// FactSpec is the specification of a Fact
+type FactSpec struct {
+	Name             string        `json:"name" protobuf:"bytes,1,opt,name=name"`
+	ID               int           `json:"id" protobuf:"bytes,2,opt,name=id"`
+	FactType         string        `json:"factType" protobuf:"bytes,3,opt,name=factType"`
+	Measurements     []Measurement `json:"measurements" protobuf:"bytes,4,opt,name=measurements"`
+	Statements       []Statement   `json:"statements" protobuf:"bytes,5,opt,name=statements"`
```

Comment:
Lack of `,omitempty` here caused issues for existing records after #2931. We should ensure that all optional fields are treated that way by the schema.

Reply:
@pmuir pointed me to [these docs](https://jenkins-x.io/contribute/apidocs/#openapi) suggesting

```golang
// +optional
```

Reply:
I can't find the difference between `optional` and `omitempty`. Shouldn't `omitempty` be enough?

Reply:
`omitempty` has an impact on the marshaled JSON, `optional` only has an impact on the schema and docs.